### PR TITLE
more spec updates

### PR DESCRIPTION
**Pull request by jlevon**
Changes include:
 - reply message includes the command number
 - split out message definitions into request/reply sections, and
   skip the repeated standard header definitions
 - lots of markup fixes
 - re-organization for clarity
 - further documentation of argsz
 - remove VFIO_USER_VM_INTERRUPT until we have a working implementation
 - dirty page tracking is optional
 - fix implementations to match the spec

Signed-off-by: John Levon <john.levon@nutanix.com>

### CODE REVIEW

**Comment by jlevon** [comment]
NB: I still have to merge in #492.

This is probably easiest to review via https://github.com/nutanix/libvfio-user/blob/more-spec-updates/docs/vfio-user.rst

**Comment by jlevon** [comment]
Fixes #486

**Comment by jlevon** [comment]
Fixes #487

File: test/py/test_device_get_region_info.py [hunk 1]
```diff
@@ -108,7 +108,7 @@ def test_device_get_region_info_larger_argsz():
     vfu_run_ctx(ctx)
     result = get_reply(sock)
 
-    assert(len(result) == 32 + 8)
+    assert(len(result) == 32)
```
Comment (tmakatos) [inline test/py/test_device_get_region_info.py:111]: Why?

**Review by tmakatos — COMMENTED** [review]
looks good in general, minor comments

File: docs/vfio-user.rst [hunk 2]
```diff
@@ -697,70 +656,69 @@ VFIO bitmap format
...
+|             | | 1   | VFIO_DEVICE_FLAGS_PCI   | |
+|             | +-----+-------------------------+ |
++-------------+--------+--------------------------+
+| num_regions | 8      | 4                        |
++-------------+--------+--------------------------+
+| num_irqs    | 12     | 4                        |
++-------------+--------+--------------------------+
+
+* *argsz* is the maximum size of the reply payload
+* all other fields must be zero.
```
[the remark this answers is not in the document]
Comment (jlevon) [reply docs/vfio-user.rst:711]: If we don't we are very out of luck if we ever want them to mean anything. There's a risk that "0" isn't valid either, but it's generally good practice IME to not just accept garbage when we have a good level of control over both sides.

File: docs/vfio-user.rst [hunk 3]
```diff
@@ -1444,310 +1218,277 @@ VFIO IRQ set format
...
 
-+-----------+--------+------+
-| Name      | Offset | Size |
-+===========+========+======+
-| Sub-index | 16     | 4    |
-+-----------+--------+------+
+``VFIO_USER_DMA_WRITE``
+-----------------------
+
+If the client has not shared mappable memory, the server can use this message to
```
[the remark this answers is not in the document]
Comment (jlevon) [reply docs/vfio-user.rst:1438]: good idea, added a caveat

File: test/py/test_device_get_region_info.py (hunk 1 above)
Comment (jlevon) [reply test/py/test_device_get_region_info.py:111]: We don't reply with the requested argsz if we have less data to report back

File: docs/vfio-user.rst [hunk 4]
```diff
@@ -697,70 +656,69 @@ VFIO bitmap format
...
 +--------+--------+------+
-| data   | 16     | 8    |
-+--------+--------+------+
 
 * *pgsize* is the page size for the bitmap, in bytes.
 * *size* is the size for the bitmap, in bytes, excluding the VFIO bitmap header.
-* *data* This field is unused in vfio-user.
-
-
-.. Note::
```
[the remark this answers is not in the document]
Comment (jlevon) [reply docs/vfio-user.rst:708]: I've lost track of what this is referring to

[the remark this answers is not in the document]
Comment (tmakatos) [reply docs/vfio-user.rst:708]: The `size` field in the VFIO bitmap message can be larger than required, e.g. if the DMA region is 8 pages then the size should be 1 byte, however QEMU sets it to 8 bytes (I learned this the hard way). So I think it's worth document this.

**Review by tmakatos — APPROVED** [review]

**Comment by jlevon** [comment]
github gone wrong

**Pull request closed without merging**